opte/src/engine/icmpv6.rs

```diff
@@ -172,3 +176,296 @@ impl HairpinAction for Icmpv6EchoReply {
+                Protocol::ICMPv6,
+            )]),
+        ];
+
+        let data_preds = vec![
+            // This must be an actual Router Solicitation message
+            DataPredicate::Icmpv6MsgType(MessageType::from(
+                Icmpv6Message::RouterSolicit,
+            )),
+        ];
+
+        (hdr_preds, data_preds)
+    }
+
+    fn gen_packet(
+        &self,
+        meta: &PacketMeta,
+        rdr: &mut PacketReader<Parsed, ()>,
+    ) -> GenPacketResult {
+        use smoltcp::time::Duration;
+        use smoltcp::wire::NdiscRouterFlags;
+        use smoltcp::wire::RawHardwareAddress;
+
+        // Collect the src / dst IP addresses, which are needed to emit the
+        // resulting ICMPv6 packet using `smoltcp`.
+        let (src_ip, dst_ip) = if let Some(metadata) = meta.inner_ip6() {
+            (
+                IpAddress::Ipv6(Ipv6Address(metadata.src.bytes())),
+                IpAddress::Ipv6(Ipv6Address(metadata.dst.bytes())),
+            )
+        } else {
+            // Getting here implies the predicate matched, but that the
+            // extracted metadata indicates this isn't an IPv6 packet. That
+            // should be impossible, but we avoid panicking given the kernel
+            // context.
+            return Err(GenErr::Unexpected(format!(
+                "Expected IPv6 packet metadata, but found: {:?}",
+                meta
+            )));
+        };
+        let body = rdr.copy_remaining();
+        let src_pkt = Icmpv6Packet::new_checked(&body)?;
+        let src_ndisc =
+            Icmpv6Repr::parse(&src_ip, &dst_ip, &src_pkt, &Csum::ignored())?;
+
+        if !matches!(
+            src_ndisc,
+            Icmpv6Repr::Ndisc(NdiscRepr::RouterSolicit { .. })
+        ) {
+            // We should never hit this case because the predicate
+            // should have verified that we are dealing with an
+            // Router Solicitation. However, programming error could
+            // cause this to happen -- let's not take any chances.
+            return Err(GenErr::Unexpected(format!(
+                "expected a NDP Router Solicitation, got {} {}",
+                src_pkt.msg_type(),
+                src_pkt.msg_code()
+            )));
+        }
+
+        // RFC 4861 6.1.1 describes a number of validation steps routers are
+        // required to perform.
+        if src_pkt.msg_code() != 0 {
+            return Ok(AllowOrDeny::Deny);
+        }
+        if !src_pkt.verify_checksum(&src_ip, &dst_ip) {
+            return Ok(AllowOrDeny::Deny);
+        }
+        // NOTE: The router is required to check that there is no Link-Layer
+        // Address Option, if the solicitation is sent from the unspecified
+        // address. The predicates here are that the solicitation came from the
+        // guest's EUI-64 transform link-local, so that is checked.
+        //
+        // TODO-completeness: Check IP Hop Limit and ICMP length / option length
```

Comment:
I might have misunderstood, but I thought by the time we get here in the code, the IPv6 header has been parsed and extracted from the `Packet`. That is, we don't really have access to the bytes of the header anymore. Is that incorrect? I also thought it might be better to avoid bloating this PR by adding the hop limit to the IPv6 header, but that could be worth it. Thoughts?

Reply:
Oh right, I'm sorry I was thinking that these are fields in the ICMP body, which is obviously not correct. You are correct that this stuff is not available here, and not really generally available at the moment. I agree skipping it in this PR is the right choice as it opens up a bigger can of worms around meta vs headers, and whether we should instead just start passing the entire headers around (since we copy them anyways) to allow for more complete inspection of a given packet. Something worth chatting about a bit next time we talk.

Reply:
Sounds good. I'll ignore this fow now, and I'm up for talking more about it next week.